Class/Mage/Buffs/MageArmorBuff.cpp

```diff
@@ -0,0 +1,24 @@
+#include "MageArmorBuff.h"
+
+#include "Spells/Buff.h"
+#include "Mage.h"
+#include "Mana.h"
+#include "Utils/Check.h"
+
+MageArmorBuff::MageArmorBuff(Mage* mage) :
+    SelfBuff(mage, "Mage Armor", "Assets/spell/Spell_arcane_mage_armor.png", BuffDuration::PERMANENT, 0), mana(static_cast<Mana*>(mage->get_resource())) {
+    check((mana != nullptr), "MageArmorBuff:: Mana nullptr");
+}
+
+void MageArmorBuff::buff_effect_when_applied() {
+    mana->mp5_from_spirit_within_5sr_modifier += 0.3;
+}
+
+void MageArmorBuff::buff_effect_when_removed() {
+    if(mana->mp5_from_spirit_within_5sr_modifier < 0.3){
```

Comment:
I would expect this entire function to simply be 

````
{
    mana->mp5_from_spirit_within_5sr_modifier -= 0.3;
}
````

What there a particular reason you saw it necessary to make sure the function is floored? If it is floating point behavior where the number can be slightly negative you can do a check using the double comparison functions in `Utils/CompareDouble.h` to assert that the modifier is never *significantly* negative instead. The problem I find with the current implementation is that it looks like it's OK if the modifier for some reason has gone much below zero, and it isn't.

Reply:
Ah my intent was not to prevent rounding errors here, but rather to prevent malfunction in case for some reason the manareg modifier was modified by some other source. E.g. you activate mage armor, then get a manareg debuff (for some reason, either on purpose or due to code errors), and then you deactivate it, then it would put you into the negative.

So I guess I could remove this if-clause if we know this will not happen, or maybe raise an exception if it's significantly negative? Same with activating mage armor, if you already have 90% manareg, you should just go to 100% right?

Reply:
I will modify it to just decrease the modifier now then. I suppose making sure that the modifier is in the correct range and not modified inappropriately is not really the job of the MargeArmor buff anyway.